src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            if (connection == null)
+            {
+                throw new ArgumentNullException(nameof(connection));
+            }
+            if (messageHandler == null)
+            {
+                throw new ArgumentNullException(nameof(messageHandler));
+            }
+            if (requestTimeout <= TimeSpan.Zero)
+            {
+                throw new ArgumentException("Doesn't expect a zero or negative timeout.", nameof(requestTimeout));
+            }
+
+            this.httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
+                Timeout = requestTimeout
+            };
+            this.httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderFactory.Create(connection);
+        }
+
+        public void Dispose()
+        {
+            if (isDisposed)
+            {
+                return;
+            }
+
+            this.httpClient.Dispose();
+            this.isDisposed = true;
+        }
+
+        public Task<Result<ComponentResponse[]>> GetComponentsSearchProjectsAsync(ComponentRequest request,
+            CancellationToken token)
+        {
+            // Since 6.2; internal
+            var apiPath = BuildRelativeUrl("api/components/search_projects",
+                new Dictionary<string, string>
+                {
+                    ["organization"] = request.OrganizationKey,
+                    ["p"] = request.Page.ToString(),
+                    ["ps"] = request.PageSize.ToString(),
+                    ["asc"] = "true",
+                });
+
+            return InvokeSonarQubeApi(apiPath, token,
+                stringResponse => JObject.Parse(stringResponse)["components"].ToObject<ComponentResponse[]>());
+        }
+
+        public Task<Result<ServerIssue[]>> GetIssuesAsync(string key, CancellationToken token)
+        {
+            // Since 5.1; internal
+            var apiPath = BuildRelativeUrl("batch/issues",
+                new Dictionary<string, string>
+                {
+                    ["key"] = key
+                });
+
+            return InvokeSonarQubeApi(apiPath, token,
+                async (HttpResponseMessage response) =>
+                {
+                    var byteArray = await response.Content.ReadAsByteArrayAsync();
+                    // Protobuf for C# throws when trying to read outside of the buffer and ReadAsStreamAsync returns a non
+                    // seekable stream so we can't determine when to stop. The hack is to use an intermediate MemoryStream
+                    // so we can control when to stop reading.
+                    // Note we might want to use FileStream instead to avoid intensive memory usage.
+                    using (var stream = new MemoryStream(byteArray))
+                    {
+                        return ReadFromProtobufStream(stream, ServerIssue.Parser).ToArray();
+                    }
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)